clang/runtime/dpct-rt/include/dpl_extras/memory.h

```diff
@@ -917,34 +928,56 @@ device_pointer<T> malloc_device(const std::size_t num_elements) {
 static inline device_pointer<void> malloc_device(const std::size_t num_bytes) {
   return device_pointer<void>(num_bytes);
 }
-template <typename T>
-device_pointer<T> device_new(device_pointer<T> p, const T &value,
+
+#ifdef DPCT_USM_LEVEL_NONE
+template <typename T, typename U>
+device_pointer<T> device_new(device_pointer<U> p, const T &value,
+                             const std::size_t count) {
+  auto converted_buf = p.buffer.template reinterpret<T>(sycl::range<1>(count));
+  ::std::uninitialized_fill(
+      oneapi::dpl::execution::make_device_policy(dpct::get_default_queue()),
+      oneapi::dpl::begin(converted_buf), oneapi::dpl::end(converted_buf),
+      value);
+  return device_pointer<T>(converted_buf, p.idx);
+}
+// buffer manages lifetime
+template <typename T> void free_device(device_pointer<T> ptr) {}
```

Comment:
I know this is just moving existing code, not new code, but as we are improving this I wonder if we want some sort of warning to the user here that for USM_NONE mode this does not actually control the "freeing".  I think the code is the right approach, but I wonder if this could get users into trouble...

@tomflinda do you have an opinion here, and what would the best way be to make such a warning?

Reply:
I'm OK to current solution.
BTW, I found a new issue for device_pointer, will file a new ticket to you.